### PR TITLE
Mattermost Connector

### LINKED ISSUE #673: Create a Mattermost connector module
Let's add a [Mattermost connector](https://mattermost.com/).

## Steps

 - Make a new submodule directory in [`opsdroid.connector`](https://github.com/opsdroid/opsdroid/tree/master/opsdroid/connector) called `mattermost` and create an `__init__.py` file inside.
 - Implement the connector according to the [documentation on writing connectors](https://opsdroid.readthedocs.io/en/stable/extending/connectors/).
 - Update the [`requirements.txt`](https://github.com/opsdroid/opsdroid/blob/master/requirements.txt) with any new dependencies from the connector if necessary.
 - Write tests for the connector. (See the [Slack connector tests](https://github.com/jacobtomlinson/opsdroid/blob/master/tests/test_connector_slack.py) for inspiration).
 - Create a [new documentation page](https://github.com/opsdroid/opsdroid/tree/master/docs/connectors).
 - Add the new page to the [mkdocs.yml](https://github.com/opsdroid/opsdroid/blob/master/mkdocs.yml).
 - Add to the [list of connectors](https://github.com/opsdroid/opsdroid/blob/master/docs/configuration-reference.md#connector-modules).

### Resources to get you started
   - [Mattermost API](https://api.mattermost.com/)
   - [Asyncio Mattermost Python module](https://pypi.org/project/mattermostdriver-asyncio/)

### Asyncio
 Opsdroid uses [asyncio and an event loop](https://docs.python.org/3/library/asyncio.html), so here are some things to bear in mind:
   - Functions should be coroutines which are defined with `async def myfunc(): ...`
   - Coroutines are called with `await`. For example `await myfunc()`. This is how asyncio allows many functions to take it in turns as they swap over when the current thread calls `await`.
   - **Don't block the thread!** If you need to make a call to a blocking function such as `time.sleep()` or any kind of IO request such as `requests.get()` then you should `await` an async variant. For example  you can call `await asyncio.sleep(10)` to sleep without blocking the whole application and you can use `aiohttp` as an async url library.

### PR DESCRIPTION
**Pull request by daniccan**
# Description

Added a new opsdroid connector for mattermost.

Fixes #673 

## Status
**READY**

## Type of change
- New feature (non-breaking change which adds functionality)
- This change requires a documentation update
- Documentation (fix or adds documentation)

# How Has This Been Tested?

- Run a sample chatbot with the following config:

```
welcome-message: true

connectors:
  - name: mattermost
    api-token: "5n45eqeh1pnd9p3t87g4ghrmic"
    url: "localhost"
    team-name: "Development"
    scheme: "http"

skills:
  - name: hello
```

- Unit test cases run by `tox`

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by daniccan** [comment]
**Pending Tasks:**

1. Fix Websocket Issue to process messages by the User.
2. Unit Test Cases for Mattermost Connector.

File: opsdroid/connector/mattermost/__init__.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
+"""A connector for Mattermost."""
+import logging
+
+from mattermostdriver import Driver
```
Comment (jacobtomlinson) [inline opsdroid/connector/mattermost/__init__.py]:
```suggestion
from mattermostdriver import Driver, Websocket
```

File: opsdroid/connector/mattermost/__init__.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            _LOGGER.info(login_response)
+
+            if "id" in login_response:
+                self.bot_id = login_response["id"]
+            if "username" in login_response:
+                self.bot_name = login_response["username"]
+
+            _LOGGER.info(_("Connected as %s"), self.bot_name)
+
+            self.mm_driver.init_websocket(self.process_message)
```
Comment (jacobtomlinson) [inline opsdroid/connector/mattermost/__init__.py]:
```suggestion
            self.mm_driver.websocket = Websocket(self.mm_driver.options, self.mm_driver.client.token)
```

File: opsdroid/connector/mattermost/__init__.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            await self.disconnect()
+            raise
+
+    async def disconnect(self):
+        """Disconnect from Mattermost."""
+        self.listening = False
+        self.mm_driver.logout()
+
+    async def listen(self):
+        """Listen for and parse new messages."""
```
Comment (jacobtomlinson) [inline opsdroid/connector/mattermost/__init__.py]:
```suggestion
        """Listen for and parse new messages."""
            await self.mm_driver.websocket.connect(self.process_message)
```

**Comment by jacobtomlinson** [comment]
Did these changes get you unstuck here?

**Comment by daniccan** [comment]
yes. thanks for the help.

**Comment by daniccan** [comment]
How to mock this function `self.mm_driver.channels.get_channel_by_name_and_team_name` for test cases ?

**Pull request merged**